diff --git a/drivers/hid/hid-nanosic-wn8030.c b/drivers/hid/hid-nanosic-wn8030.c
index 37b353ede5f4..3a056c971c43 100644
--- a/drivers/hid/hid-nanosic-wn8030.c
+++ b/drivers/hid/hid-nanosic-wn8030.c
@@ -466,6 +466,10 @@ static void nanosic_wn8030_handle_vendor(struct nanosic_wn8030 *nanosic, u8 *buf
 		}
 		mutex_unlock(&nanosic->conn_mutex);
 	}
+	/* WN8012(KB) -> HOST, kb enters sleep mode */
+	else if (buf[5] == 0x38 && buf[6] == 0x80 && buf[7] == 0x28 && buf[9] == 0x00) {
+		nanosic_wn8030_set_kb_power(nanosic, true);
+	}
 }
 
 static irqreturn_t nanosic_wn8030_handler(int irq, void *data)
@@ -500,6 +504,7 @@ static irqreturn_t nanosic_wn8030_handler(int irq, void *data)
 			hid_input_report(nanosic->hid_touchpad, HID_INPUT_REPORT,
 					 &buf[3], 21, 0);
 		break;
+	case 0x22:
 	case 0x23:
 		nanosic_wn8030_handle_vendor(nanosic, buf);
 		break;
