diff --git a/drivers/hid/hid-nanosic-wn8030.c b/drivers/hid/hid-nanosic-wn8030.c
index 37b353ede5f4..1e5ed9905f60 100644
--- a/drivers/hid/hid-nanosic-wn8030.c
+++ b/drivers/hid/hid-nanosic-wn8030.c
@@ -360,6 +360,17 @@ static int nanosic_wn8030_set_kb_power(struct nanosic_wn8030 *nanosic, bool enab
 	return regmap_bulk_write(nanosic->regmap, 0x5c, buf, sizeof(buf));
 }
 
+static int nanosic_wn8030_set_kb_sleep(struct nanosic_wn8030 *nanosic, bool enable)
+{
+	u8 buf[XM_WN8030_I2C_WRITE] = { 0x32, 0x00, 0x4E, 0x31,
+					0x80, 0x38, 0x28, 0x01 };
+
+	buf[8] = !enable;
+	buf[9] = nanosic_wn8030_checksum8(&buf[2], 7);
+
+	return regmap_bulk_write(nanosic->regmap, 0x5c, buf, sizeof(buf));
+}
+
 static int nanosic_wn8030_output_report(struct hid_device *hid, u8 *buf, size_t count)
 {
 	struct nanosic_wn8030 *nanosic = hid->driver_data;
@@ -466,6 +477,10 @@ static void nanosic_wn8030_handle_vendor(struct nanosic_wn8030 *nanosic, u8 *buf
 		}
 		mutex_unlock(&nanosic->conn_mutex);
 	}
+	/* WN8012(KB) -> HOST, kb enters sleep mode */
+	else if (buf[5] == 0x38 && buf[6] == 0x80 && buf[7] == 0x28 && buf[9] == 0x00) {
+		nanosic_wn8030_set_kb_sleep(nanosic, false);
+	}
 }
 
 static irqreturn_t nanosic_wn8030_handler(int irq, void *data)
@@ -500,6 +515,7 @@ static irqreturn_t nanosic_wn8030_handler(int irq, void *data)
 			hid_input_report(nanosic->hid_touchpad, HID_INPUT_REPORT,
 					 &buf[3], 21, 0);
 		break;
+	case 0x22:
 	case 0x23:
 		nanosic_wn8030_handle_vendor(nanosic, buf);
 		break;
